### PR TITLE
Provide Google OAuth login internally

### PR DESCRIPTION
**Pull request by Rindrics**
## Summary

Enables user to login using Google account internally using feature flag

## Related Issue

- #92 

## Changes

- add "Google" type social signup & login feature
- add Google OAuth management feature for /settings/account page

## Testing

- [x] "Sign in with Google" button is absent by default
- [x] "with Google" button is available with its feature flag on
- [x] can signup "with Google"
- [x] can login "with Google"
- [x] can connect & disconnect with Google
- [x] preexistent GitHub OAuth feature (signup, login, connect & disconnect) works

## Other Information

I did not refactor codes and simply repeated the preexistent codes for GitHub OAuth to focus to delivering the feature

### CODE REVIEW

**Comment by shige** [comment]
@satococoa Since you've implemented the GitHub login, please review this pr.

**Comment by satococoa** [comment]
@Rindrics 

Do you plan to implement the authentication settings on the `/settings/account` page in another pull request?

スクリーンショット 2024-11-15 8 59 03

File: app/(auth)/auth/callback/route.ts [hunk 1]
```diff
@@ -72,23 +73,37 @@ async function initializeUserIfNeeded(user: User) {
...
 
-	let provider = "";
-	// https://docs.github.com/ja/authentication/keeping-your-account-and-data-secure/about-authentication-to-github#github-%E3%81%AE%E3%83%88%E3%83%BC%E3%82%AF%E3%83%B3%E3%83%95%E3%82%A9%E3%83%BC%E3%83%9E%E3%83%83%E3%83%88
-	if (provider_token.startsWith("ghu_")) {
-		provider = "github";
-	}
-	// TODO: add another logic for other providers
-	if (provider === "") {
-		throw new Error("No provider found");
+	const matchedProvider = oauth_providers.find((p) => p.isMatch(session));
```
Comment (satococoa) [inline app/(auth)/auth/callback/route.ts]: In my Supabase user object, `app_metadata` is as follows:

```
app_metadata: { provider: 'email', providers: [ 'email', 'github' ] },
```

After I connecting my Google account, it may look like this:

```
app_metadata: { provider: 'email', providers: [ 'email', 'github', 'google' ] },
```

I am not sure how the order of app_metadata.providers is determined. 
However this logic will attempt to use GitHub provider's access token and refresh token, even if I've just authenticated with Google OAuth.

**Review by satococoa — CHANGES_REQUESTED** [review]

File: app/(auth)/auth/callback/route.ts (hunk 1 above)
Comment (rindrics) [reply app/(auth)/auth/callback/route.ts]: thanks, I'll correct the login logic to match user intent and backend data @satococoa

**Comment by rindrics** [comment]
> Do you plan to implement the authentication settings on the /settings/account page in another pull request?

no~~~ I forgot to implement that page, but I'll add it onto this pull request as new commits.

thank you @satococoa

File: app/(auth)/auth/callback/route.ts (hunk 1 above)
Comment (rindrics) [reply app/(auth)/auth/callback/route.ts]: I [reverted it](https://github.com/giselles-ai/giselle/pull/107/commits/23c6afe13bf20ead9577386667c93060fd19226f) and implemented "settings/" page.

would you re-review the codes? @satococoa

File: app/(auth)/signup/page.tsx [hunk 2]
```diff
@@ -24,7 +24,7 @@ export default function SignupPage() {
 			</div>
 			<div className="mt-8 space-y-6 w-[320px]">
 				<PageHeader title="Get Started" />
-				<OAuthProviders labelPrefix="Sign up" />
+				<OAuthProviders labelPrefix="Continue" />
```
Comment (satococoa) [inline app/(auth)/signup/page.tsx]: Why did you change this?

File: app/(main)/settings/account/actions.ts [hunk 3]
```diff
@@ -77,3 +78,77 @@ export async function disconnectGitHubIdentity() {
 
 	await deleteOauthCredential("github");
 }
+
+export async function connectGoogleIdentity() {
```
Comment (satococoa) [inline app/(main)/settings/account/actions.ts:83]: These methods share similar logic with GitHub's implementation, suggesting an opportunity for refactoring.

The proposed unified structure would be:

- function connectIdentity(provider: Provider)
- function reconnectIdentity(provider: Provider)
- function disconnectIdentity(provider: Provider)

File: app/(main)/settings/account/github-authentication.tsx [hunk 4]
```diff
@@ -16,7 +16,7 @@ import {
 export async function GitHubAuthentication() {
 	const credential = await getOauthCredential("github");
 	if (!credential) {
-		return <GitHubAuthentcationPresentation button={GitHubConnectButton} />;
+		return <GitHubAuthenticationPresentation button={GitHubConnectButton} />;
```
Comment (satococoa) [inline app/(main)/settings/account/github-authentication.tsx:23]: Thanks for the correction 👍 👍 
My mistake.

File: app/(auth)/auth/callback/route.ts [hunk 5]
```diff
@@ -84,14 +101,20 @@ async function storeProviderTokens(user: User, session: Session) {
 	if (provider_token.startsWith("ghu_")) {
 		provider = "github";
 	}
+	if (provider_token.startsWith("ya29.")) {
```
Comment (satococoa) [inline app/(auth)/auth/callback/route.ts]: I couldn't find any official documentation about this behavior in a quick search.

Proposal: Use path parameters to identify the auth provider

Suggested changes:
1. Modify route file structure to use dynamic provider parameter:
   - `app/(auth)/auth/callback/[provider]/route.ts`

2. Update `getAuthCallbackUrl` to generate provider-specific paths:
   - `/auth/callback/github`
   - `/auth/callback/google`

This approach aligns better with Next.js routing conventions and improves code maintainability.

Comment (satococoa) [reply app/(auth)/auth/callback/route.ts]: This approach using path parameters is more robust than my previous implementation that relied on token prefix checking (`provider_token.startsWith("ghu_")`).

File: app/(main)/settings/components/google-connection-button.tsx
Comment (satococoa) [inline app/(main)/settings/components/google-connection-button.tsx:1]: Can we refactor `GitHubConnectionButton` to `ProviderConnectionButton` and reuse it for all providers?

File: sentry.client.config.ts [hunk 6]
```diff
@@ -23,5 +23,5 @@ Sentry.init({
 
 	// Setting this option to true will print useful information to the console while you're setting up Sentry.
 	debug: false,
-	enabled: process.env.NEXT_PUBLIC_VERCEL_ENV !== undefined,
+	enabled: true,
```
Comment (satococoa) [inline sentry.client.config.ts]: Was this change intended?

File: services/external/google/user-client.ts [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+
+export type GoogleUserData = {
+	name: string;
+	email: string;
+};
+
+class GoogleUserClient {
+	private clientId: string;
+	private clientSecret: string;
+	private logger: Logger;
```
Comment (satococoa) [inline services/external/google/user-client.ts]: I've found that this `logger` is not used in this class.
If you don't need the reporting sentry feature here, consider remove this property.

**Review by satococoa — CHANGES_REQUESTED** [review]
Thank you for implementing! I've added some comments. Could you check them out?

File: app/(main)/settings/components/google-connection-button.tsx
Comment (rindrics) [reply app/(main)/settings/components/google-connection-button.tsx:1]: yes, I think so.

I'll try that in this PR @satococoa

File: app/(auth)/auth/callback/route.ts (hunk 5 above)
Comment (rindrics) [reply app/(auth)/auth/callback/route.ts]: > This approach aligns better with Next.js routing conventions and improves code maintainability.

thank you, I'll try that in this PR @satococoa

File: services/external/google/user-client.ts (hunk 7 above)
Comment (rindrics) [reply services/external/google/user-client.ts]: you're right.
I removed it by c0320c6

File: sentry.client.config.ts (hunk 6 above)
Comment (rindrics) [reply sentry.client.config.ts]: no, I'm sorry about that.
I reverted the change in 49eb1e4. thanks

File: app/(auth)/signup/page.tsx (hunk 2 above)
Comment (rindrics) [reply app/(auth)/signup/page.tsx]: no, it was my mistake introduced during pasting codes from my testing env!
I'll revert the change by de32c45

**Comment by rindrics** [comment]
Although I did not include refactoring in this PR, I'll reflect the suggested changes below:
- https://github.com/giselles-ai/giselle/pull/107#discussion_r1853220600
- https://github.com/giselles-ai/giselle/pull/107#discussion_r1853230094
- https://github.com/giselles-ai/giselle/pull/107#discussion_r1853233141

File: app/(main)/settings/account/actions.ts (hunk 3 above)
Comment (rindrics) [reply app/(main)/settings/account/actions.ts:83]: I unified the functions by the commits below:
- 6b5f484
- a427861
- 4d14660

File: app/(auth)/auth/callback/route.ts (hunk 5 above)
Comment (rindrics) [reply app/(auth)/auth/callback/route.ts]: I improved the provider handling by efd9f98

File: app/(main)/settings/components/google-connection-button.tsx
Comment (rindrics) [reply app/(main)/settings/components/google-connection-button.tsx:1]: They were completely duplicated components. thanks
fixed: b21557c

**Comment by rindrics** [comment]
> Although I did not include refactoring in this PR, I'll reflect the suggested changes below:
> 
> - https://github.com/giselles-ai/giselle/pull/107#discussion_r1853220600
> - https://github.com/giselles-ai/giselle/pull/107#discussion_r1853230094
> - https://github.com/giselles-ai/giselle/pull/107#discussion_r1853233141

I reopen this PR because the suggested refactors are all done

File: app/(main)/settings/account/google-authentication.tsx [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+				googleUser={googleUser}
+				button={unlinkable ? GoogleDisconnectButton : undefined}
+			/>
+		);
+	} catch (error) {
+		if (needsAuthorization(error)) {
+			return (
+				<GoogleAuthenticationPresentation
+					button={GoogleReconnectButton}
+					alert="Your GitHub access token has expired or become invalid. Please reconnect to continue using the service."
```
Comment (satococoa) [inline app/(main)/settings/account/google-authentication.tsx]:
```suggestion
					alert="Your Google access token has expired or become invalid. Please reconnect to continue using the service."
```

File: services/external/google/user-client.ts [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+			const authClient = await this.buildClient();
+			logger.debug(
+				{
+					hasAccessToken: !!this.token.accessToken,
+					tokenPrefix: this.token.accessToken?.substring(0, 5),
+				},
+				"getting user info",
+			);
+
+			const oauth2 = google.oauth2({
```
Comment (satococoa) [inline services/external/google/user-client.ts:78]: [nits]I think you can build the `google.oauth2` instance inside of `buildClient()`.
However, it is trivial for now, so it's okay here.

**Review by toyamarinyon — COMMENTED** [review]
Thanks! 

I tried google login and visited `settings/account` then faced the client error. reloading the browser did not reproduce it. Do you have any idea what it could be?

https://giselle-git-issue-53-r06-edge.vercel.app/settings/account

File: package.json [hunk 10]
```diff
@@ -62,6 +62,8 @@
 		"cva": "1.0.0-beta.1",
 		"drizzle-orm": "0.32.1",
 		"framer-motion": "12.0.0-alpha.0",
+		"google-auth-library": "^9.15.0",
+		"googleapis": "^144.0.0",
```
Comment (toyamarinyon) [inline package.json]: You can use `-E` option to add a package and pin to the resolved version.
https://bun.sh/docs/cli/add#exact
```suggestion
		"google-auth-library": "9.15.0",
		"googleapis": "144.0.0",
```

**Comment by rindrics** [comment]
> I tried google login and visited settings/account then faced the client error. reloading the browser did not reproduce it. Do you have any idea what it could be?

no, I'll investigate it and will fix the problem in the future commits in this PR. Thanks!

**Comment by rindrics** [comment]
> I tried google login and visited settings/account then faced the client error. reloading the browser did not reproduce it. Do you have any idea what it could be?

Some investigation revealed that the error is not introduced by this PR itself.

I created https://github.com/giselles-ai/giselle/issues/130 and noted that users can resolve the error by page reload for workaround.

Would you continue reviewing because #130 is independent of this PR? @toyamarinyon @shige

**Comment by shige** [comment]
I'll be reviewing this PR now!

**Review by shige — APPROVED** [review]
LGTM! 🆔

**Review by toyamarinyon — APPROVED** [review]

**Comment by rindrics** [comment]
thank you all for reviewing 👍
